resolwe_bio/kb/views.py

```diff
@@ -123,34 +99,77 @@ def create(self, request):
-        """Support POST for searching against a list of feature ids."""
+        """Support POST for searching against a list of genes."""
         if actions.get('get', None) == 'list':
             actions['post'] = 'list_with_post'
 
-        return super(cls, MappingViewSet).as_view(actions, **initkwargs)
+        return super(cls, MappingSearchViewSet).as_view(actions, **initkwargs)
+
+    def filter_queryset(self, queryset):
+        """Support filtering by a list of genes."""
+        queryset = super(MappingSearchViewSet, self).filter_queryset(queryset)
+        if 'source_id' in self.request.data:
+            queryset = queryset.filter(source_id__in=self.request.data['source_id'])
+        if 'source_db' in self.request.data:
+            queryset = queryset.filter(source_db=self.request.data['source_db'])
+        if 'target_id' in self.request.data:
+            queryset = queryset.filter(target_id__in=self.request.data['target_id'])
+        if 'target_db' in self.request.data:
+            queryset = queryset.filter(target_db=self.request.data['target_db'])
+        if 'relation_type' in self.request.data:
+            queryset = queryset.filter(relation_type=self.request.data['relation_type'])
+
+        return queryset
 
     def list_with_post(self, request):
         """Support search via a POST request in addition to GET."""
-        queryset = self.get_queryset()
-        filter_params = request.data
-        for key, value in filter_params.items():
-            if isinstance(value, list):
-                filter_params[key] = ','.join(value)
-
-        if filter_params:
-            queryfilter = self.filter_class(filter_params, queryset=queryset)
-            queryset = queryfilter.qs
-        serializer = self.get_serializer(queryset, many=True)
-        return Response(serializer.data)
+        return self.list(request)
+
+
+class MappingViewSet(mixins.ListModelMixin,
+                     mixins.RetrieveModelMixin,
+                     mixins.CreateModelMixin,
+                     mixins.UpdateModelMixin,
+                     mixins.DestroyModelMixin,
+                     viewsets.GenericViewSet):
+    """API view for :class:`Mapping` objects."""
+
+    serializer_class = MappingSerializer
+    permission_classes = [permissions.IsAdminUser]
+    filter_backends = [filters.DjangoFilterBackend]
+    filter_class = MappingFilter
+    queryset = Mapping.objects.all()
+
+    def create(self, request):
+        """A custom create, which updates existing mappings instead of failing."""
+        try:
+            mapping = Mapping.objects.get(
+                source_db=request.data['source_db'],
+                source_id=request.data['source_id'],
+                target_db=request.data['target_db'],
+                target_id=request.data['target_id']
+            )
+            self.kwargs[self.lookup_field] = mapping.pk
+            return super(MappingViewSet, self).update(request)  # pylint: disable=no-member
```

Comment:
Yes, we also have this for features to avoid a needless round-trip. Otherwise the API would first need to check if the object  already exists on the server and then create or update (and this would also introduce a potential race condition).